### PR TITLE
Report pulled image size

### PR DESCRIPTION
**Pull request by nievesmontero**
https://github.com/containers/toolbox/issues/752

Signed-off-by: Nieves Montero <nmontero@redhat.com>

### CODE REVIEW

**Comment by debarshiray** [comment]
recheck

**Comment by debarshiray** [comment]
The previous `RETRY_LIMIT` on Fedora 35 seems to have been a transient error:
```
TASK [Install RPM packages]
fedora-35 | ERROR
fedora-35 | {
fedora-35 |   "msg": "Failed to download metadata for repo 'fedora': Cannot prepare internal mirrorlist: Status code: 404 for https://mirrors.fedoraproject.org/metalink?repo=fedora-35&arch=x86_64 (IP: 38.145.60.21)",
fedora-35 |   "rc": 1,
fedora-35 |   "results": []
fedora-35 | }
```

File: src/pkg/podman/podman.go [hunk 1]
```diff
@@ -280,6 +282,47 @@ func Inspect(typearg string, target string) (map[string]interface{}, error) {
...
+
+	if err := shell.Run("skopeo", nil, &stdout, nil, args...); err != nil {
+		return nil, err
+	}
+
+	output := stdout.Bytes()
+
+	var info interface{}
+
+	if err := json.Unmarshal(output, &info); err != nil {
```
Comment (debarshiray) [inline src/pkg/podman/podman.go:299]: ... then we can have a custom type called `Image` that implements the [json.Unmarshaler](https://pkg.go.dev/encoding/json#Unmarshaler) interface.  [This](https://go.dev/blog/json) gives a good introduction to handling JSON in Go.

There's a similar `Image` type in `src/pkg/podman/podman.go` that's used to parse JSON in `GetImages()`.  We could have used this same type to parse Skopeo's JSON, but it's different from Podman's JSON, so it might be better to keep them separate.

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for working on this, Nieves!  Some comments below:

File: src/cmd/create.go [hunk 2]
```diff
@@ -646,6 +647,7 @@ func getServiceSocket(serviceName string, unitName string) (string, error) {
 }
 
 func pullImage(image, release string) (bool, error) {
+	logrus.Debugf("This is the image we're looking for %s", image)
```
Comment (debarshiray) [inline src/cmd/create.go:650]: Nitpick: "Looking for image %s" will be better because it matches the logs elsewhere.

File: src/cmd/create.go [hunk 3]
```diff
@@ -666,18 +668,19 @@ func pullImage(image, release string) (bool, error) {
 	}
 
 	var imageFull string
+	var imageSize string = "docker://"
```
Comment (debarshiray) [inline src/cmd/create.go:671]: The `docker://` prefix looks like an implementation detail of the `skopeo inspect ...` wrapper.  So, I would suggest moving it there.

**Review by debarshiray — CHANGES_REQUESTED** [review]

**Comment by debarshiray** [comment]
Due to the recent expansion of the test suite, we might have to increase the timeout for stable Fedoras: https://github.com/containers/toolbox/pull/1212

We recently had to do that for Fedora Rawhide.

File: src/cmd/create.go [hunk 4]
```diff
@@ -554,6 +554,7 @@ func getFullyQualifiedImageFromRepoTags(image string) (string, error) {
 		imageFull = image
 	} else {
 		info, err := podman.Inspect("image", image)
+
```
[the remark this answers is not in the document]
Comment (debarshiray) [reply src/cmd/create.go:557]: Unresolving, because I still see the stray newline in the latest version.

File: src/cmd/create.go (hunk 2 above)
Comment (debarshiray) [reply src/cmd/create.go:650]: Unresolving, because the log message is still the same.

File: src/cmd/create.go [hunk 5]
```diff
@@ -666,18 +668,19 @@ func pullImage(image, release string) (bool, error) {
 	}
 
 	var imageFull string
+	var imageSize string = "docker://"
 
 	if hasDomain {
 		imageFull = image
+
```
[the remark this answers is not in the document]
Comment (debarshiray) [reply src/cmd/create.go:675]: Unresolving, because I still see the stray newline in the latest version.

File: src/pkg/podman/podman.go [hunk 6]
```diff
@@ -261,6 +261,7 @@ func ImageExists(image string) (bool, error) {
 //
 // Parameter 'typearg' takes in values 'container' or 'image' that is passed to the --type flag
 func Inspect(typearg string, target string) (map[string]interface{}, error) {
+
```
Comment (debarshiray) [inline src/pkg/podman/podman.go:264]: Nitpick: stray newline.

File: src/pkg/podman/podman.go [hunk 7]
```diff
@@ -271,6 +272,7 @@ func Inspect(typearg string, target string) (map[string]interface{}, error) {
 	}
 
 	output := stdout.Bytes()
+
```
Comment (debarshiray) [inline src/pkg/podman/podman.go:275]: Ditto.

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for rebasing this on top of https://github.com/containers/toolbox/pull/1212

We still need to address the review comments about the code, though.

**Comment by debarshiray** [comment]
recheck

File: src/pkg/skopeo/skopeo.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+package skopeo
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/containers/toolbox/pkg/shell"
+)
+
+func Inspect(typearg string, target string) (interface{}, error) {
```
Comment (debarshiray) [inline src/pkg/skopeo/skopeo.go:26]: The `typearg` argument is unused, so it might be better to not have it.

The reason `podman.Inspect` has a `typearg` argument is because one can use `podman inspect` with both OCI containers and images.  However, that's not the case with `skopeo inspect` because it can only be used with images.

File: src/pkg/skopeo/skopeo.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright © 2019 – 2022 Red Hat Inc.
```
Comment (debarshiray) [inline src/pkg/skopeo/skopeo.go:2]: Nitpick: should be `Copyright © 2023 Red Hat Inc.`.

File: src/cmd/create.go [hunk 10]
```diff
@@ -698,9 +699,15 @@ func pullImage(image, release string) (bool, error) {
 	}
 
 	if promptForDownload {
+
+		infoSize, err := podman.Inspect("image", imageSize)
```
Comment (debarshiray) [inline src/cmd/create.go:703]: Shouldn't it be `skopeo.Inspect`, not `podman.Inspect`?

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for the updates, @nievesmontero !  Could you please squash the commits into one? You can use git rebase -i for that.

Further comments below:

File: src/pkg/podman/podman.go [hunk 11]
```diff
@@ -280,6 +282,47 @@ func Inspect(typearg string, target string) (map[string]interface{}, error) {
 	return info[0], nil
 }
 
+func InspectSkopeo(typearg string, target string) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (debarshiray) [reply src/pkg/podman/podman.go]: This has been done in the latest iteration of changes.

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for the updates, @nievesmontero !  It looks like there are some remnants from a `git rebase` or `git merge` or something similar that had conflicts and wasn't fully cleaned up.

File: src/cmd/create.go [hunk 12]
```diff
@@ -646,6 +647,7 @@ func getServiceSocket(serviceName string, unitName string) (string, error) {
 }
 
 func pullImage(image, release string) (bool, error) {
+	logrus.Debugf("Looking for image %s", image)
```
Comment (debarshiray) [inline src/cmd/create.go:650]: Nitpick: a newline after the `logrus.Debugf` will be good because it will separate the log message from the actual implementation of the function.

**Review by debarshiray — CHANGES_REQUESTED** [review]
Some of the comments from the past reviews still need to be addressed.

**Review by debarshiray — CHANGES_REQUESTED** [review]

**Comment by debarshiray** [comment]
The main things that are left to do here are:

* Avoid a hard dependency on Skopeo by ensuring that the errors from `skopeo inspect ...` are not hard failures.

* Use a custom `Image` type that implements the [json.Unmarshaler](https://pkg.go.dev/encoding/json#Unmarshaler) interface for parsing JSON.

* Move the `docker://` prefix into `skopeo.Inspect`.

See above for details on each of these points.

File: src/pkg/podman/podman.go [hunk 13]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2019 – 2022 Red Hat Inc.
+ * Copyright © 2023 Red Hat Inc.
```
[the remark this answers is not in the document]
Comment (nievesmontero) [reply src/pkg/podman/podman.go:2]: In one of the previous comments you told me it should be 2023 instead of 2019-2022 https://github.com/containers/toolbox/pull/1208#discussion_r1087708243
Should I leave it then just the way it was?

[the remark this answers is not in the document]
Comment (debarshiray) [reply src/pkg/podman/podman.go:2]: Yes, that was for `src/pkg/skopeo/skopeo.go`, but this is `src/pkg/podman/podman.go`.

Usually, you only update the copyright notices in the files that you touch.

[the remark this answers is not in the document]
Comment (nievesmontero) [reply src/pkg/podman/podman.go:2]: My fault, I didn't read the file correctly, thank you!

File: src/cmd/create.go [hunk 14]
```diff
@@ -699,9 +699,14 @@ func pullImage(image, release, authFile string) (bool, error) {
 	}
 
 	if promptForDownload {
+		infoSize, err := skopeo.Inspect(imageFull)
```
Comment (debarshiray) [inline src/cmd/create.go:702]: I see that you are now passing `imageFull`, which doesn't have the `docker://` prefix.  That's better.  However, I don't see the prefix being added inside `skopeo.Inspect`.

File: src/pkg/podman/podman.go [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © 2019 – 2022 Red Hat Inc.
+ * Copyright © 2019-2022 Red Hat Inc.
```
Comment (debarshiray) [inline src/pkg/podman/podman.go:2]: Stray unrelated change.

**Review by debarshiray — CHANGES_REQUESTED** [review]

File: src/pkg/podman/podman.go (hunk 15 above)
Comment (nievesmontero) [reply src/pkg/podman/podman.go:2]: This change was made by mistake after fixing some problems I had.

File: src/cmd/create.go (hunk 14 above)
Comment (nievesmontero) [reply src/cmd/create.go]: Now I am passing the string ```docker://``` inside the ```skopeo.Inspect()``` method, I hope that works better.

File: src/pkg/podman/podman.go (hunk 15 above)
Comment (debarshiray) [reply src/pkg/podman/podman.go]: Unresolving, because it's still there.  :)

File: src/cmd/create.go [hunk 16]
```diff
@@ -678,8 +679,6 @@ func pullImage(image, release, authFile string) (bool, error) {
 		}
 	}
 
-	logrus.Debugf("Looking for image %s", imageFull)
-
```
Comment (debarshiray) [inline src/cmd/create.go:682]: This debug message shouldn't be removed.

**Review by debarshiray — CHANGES_REQUESTED** [review]

File: src/cmd/create.go (hunk 16 above)
Comment (nievesmontero) [reply src/cmd/create.go]: Included again :)

File: src/pkg/podman/podman.go (hunk 15 above)
Comment (nievesmontero) [reply src/pkg/podman/podman.go]: Modified :)

File: src/cmd/create.go (hunk 14 above)
Comment (debarshiray) [reply src/cmd/create.go]: No, it should be inside the body of the `skopeo.Inspect` function.

Like I said [before](https://github.com/containers/toolbox/pull/1208#discussion_r1073452041), the `docker://` prefix is something that `skopeo inspect` needs.  The caller of `skopeo.Inspect` doesn't need it and doesn't need to know about it.  So, it should be inside `skopeo.Inspect`.

**Comment by nievesmontero** [comment]
recheck

File: src/cmd/create.go [hunk 17]
```diff
@@ -699,9 +700,14 @@ func pullImage(image, release, authFile string) (bool, error) {
 	}
 
 	if promptForDownload {
+		infoSize, err := skopeo.Inspect(imageFull)
+		if err != nil {
+			logrus.Debug("cannot get the size of the image")
```
Comment (debarshiray) [inline src/cmd/create.go:705]: Since we are suppressing the original error, it will be good to dump it in the logs.  Something like:
```
  logrus.Debugf("Getting image size failed: %s", err)
```

File: src/cmd/create.go [hunk 18]
```diff
@@ -699,9 +700,14 @@ func pullImage(image, release, authFile string) (bool, error) {
...
+		}
+
 		fmt.Println("Image required to create toolbox container.")
 
-		prompt := fmt.Sprintf("Download %s (500MB)? [y/N]:", imageFull)
+		prompt := fmt.Sprintf("Download %s (%vMB)? [y/N]:", imageFull, infoSize)
```
Comment (debarshiray) [inline src/cmd/create.go:710]: If we fail to get the image size, then this message will look odd:
```
Download <fully-qualified-image-name> [0MB]? [y/N]:
```

In many cases, it will fail because the `skopeo(1)` binary was absent from the system, which I think should be a [valid](https://github.com/containers/toolbox/issues/752) use-case.  So we should try to cover those situations elegantly.

**Review by debarshiray — CHANGES_REQUESTED** [review]
Thanks for the updates, @nievesmontero !  This is looking a lot better, well done.  Some minor details:

**Review by debarshiray — COMMENTED** [review]
One quick note about the commit message ...

Unless the change is massive, it's not so important to write things like:
```
The following package and file have been created src/pkg/skopeo/skopeo.go
```

... because it's obvious from looking at the commit (with `git show`) and is of no interest to downstream distributors.  What's more important is that it adds an optional dependency on Skopeo, because that's going to be important for downstream distributors, and to clarify why it's optional.  Those details aren't so obvious until someone carefully looks at the commit, but they are still important.

File: src/cmd/create.go (hunk 18 above)
Comment (nievesmontero) [reply src/cmd/create.go:710]: I tried to solve this by adding an else on the code. Check if it works for you or if it could be done in a more elegantly way.

File: src/cmd/create.go [hunk 19]
```diff
@@ -699,10 +700,20 @@ func pullImage(image, release, authFile string) (bool, error) {
...
+			fmt.Println("Image required to create toolbox container.")
 
-		prompt := fmt.Sprintf("Download %s (500MB)? [y/N]:", imageFull)
-		shouldPullImage = askForConfirmation(prompt)
+			prompt := fmt.Sprintf("Download %s (500MB)? [y/N]:", imageFull)
+			shouldPullImage = askForConfirmation(prompt)
+		} else {
+			fmt.Println("Image required to create toolbox container.")
+
+			prompt := fmt.Sprintf("Download %s (%sMB)? [y/N]:", imageFull, infoSize)
```
Comment (debarshiray) [inline src/cmd/create.go:714]: The `MB` is getting duplicated because it's part of the human-readable size.  :)

File: src/pkg/skopeo/skopeo.go [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+	var image Image
+	if err := json.Unmarshal(output, &image); err != nil {
+		return "", err
+	}
+
+	var totalSize int64 = 0
+	for _, layer := range image.LayersData {
+		totalSize += layer.Size
+	}
+	sizeInMB := units.HumanSize(float64(totalSize))
```
Comment (debarshiray) [inline src/pkg/skopeo/skopeo.go:53]: Nitpick: see above.

The size might not necessarily be in MB.  It's just in a human-readable form.  eg., It could also be in GB for sufficiently large images.

**Review by debarshiray — CHANGES_REQUESTED** [review]

File: src/pkg/skopeo/skopeo.go (hunk 20 above)
Comment (nievesmontero) [reply src/pkg/skopeo/skopeo.go:53]: You're right

File: src/pkg/skopeo/skopeo.go [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+		return "", err
+	}
+
+	output := stdout.Bytes()
+	var image Image
+	if err := json.Unmarshal(output, &image); err != nil {
+		return "", err
+	}
+
+	var totalSize int64 = 0
```
Comment (debarshiray) [inline src/pkg/skopeo/skopeo.go:49]: Nitpick: just `var totalSize int64` is enough.  There's no need for the `= 0`.

File: src/pkg/skopeo/skopeo.go [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+	targetWithTransport := "docker://" + target
+
+	args := []string{"inspect", targetWithTransport}
+	if err := shell.Run("skopeo", nil, &stdout, nil, args...); err != nil {
+		return "", err
+	}
+
+	output := stdout.Bytes()
+	var image Image
+	if err := json.Unmarshal(output, &image); err != nil {
```
Comment (debarshiray) [inline src/pkg/skopeo/skopeo.go:45]: Like we were discussing on chat, were you able to test the size calculation with a Skopeo older than `v1.10.0`?  That's when the LayersData array was [added](https://github.com/containers/skopeo/pull/1738) to `skopeo inspect` so, I am curious how it handles older versions.

Comment (debarshiray) [reply src/pkg/skopeo/skopeo.go:45]: I tested with an older Skopeo and found that the prompt was:
```
Download <fully-qualified-image-name> [0B]? [y/N]:
```

I took the liberty to make the necessary changes to not show that `0B`, as we discussed [before](https://github.com/containers/toolbox/pull/1208#discussion_r1126577835).
